Update singleton int to error when inequality relation is undefined

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110529
* #110528
* #109171
* #110372
* #110371
* #110370
* #110369
* __->__ #110044

Previously, something like j0 >= 3, would return False. In sympy however, it is not possible to make it so that both j0 >= 3 and j0 < 3 return False. In sympy, you only get to dispatch on Ge, and the remaining are derived, e.g. defining Ge(j0 >= 3) to be False would force Lt(j0, 3) to be True, which is not what we want. 

In this PR, we make it so that both j0 >=3 and j0 < 3 error, so that in a future PR when we create the symbolic counterpart of this singleton, the behaviors can be the same.

soulitzer committed:
Support multiplication with scalar factor for SingletonSymNode

[ghstack-poisoned]

soulitzer committed:
Update on "Support multiplication with scalar factor for SingletonSymNode"




[ghstack-poisoned]

soulitzer committed:
Update on "Support multiplication with scalar factor for SingletonSymNode"


Support multiplication with scalars allows us to support strides. A naive alternative is just to create an entirely new singleton whenever you need a new stride, but that doesn't work because it is important during tracing that the strides of the outputs can be computed using the size/strides of the inputs. 

[ghstack-poisoned]

soulitzer committed:
Update on "Update singleton int to error when inequality relation is undefined"


Previously, something like j0 >= 3, would return False. In sympy however, it is not possible to make it so that both j0 >= 3 and j0 < 3 return False. In sympy, you only get to dispatch on Ge, and the remaining are derived, e.g. defining Ge(j0 >= 3) to be False would force Lt(j0, 3) to be True, which is not what we want. 

In this PR, we make it so that both j0 >=3 and j0 < 3 error, so that in a future PR when we create the symbolic counterpart of this singleton, the behaviors can be the same.

[ghstack-poisoned]

soulitzer committed:
Update on "Update singleton int to error when inequality relation is undefined"


Previously, something like j0 >= 3, would return False. In sympy however, it is not possible to make it so that both j0 >= 3 and j0 < 3 return False. In sympy, you only get to dispatch on Ge, and the remaining are derived, e.g. defining Ge(j0 >= 3) to be False would force Lt(j0, 3) to be True, which is not what we want. 

In this PR, we make it so that both j0 >=3 and j0 < 3 error, so that in a future PR when we create the symbolic counterpart of this singleton, the behaviors can be the same.

[ghstack-poisoned]

ezyang commented on c10/core/SingletonSymNodeImpl.h:
```diff
@@ -81,42 +108,72 @@ class C10_API SingletonSymNodeImpl : public SymNodeImpl {
-    if (auto mb_si = other->singleton_int()) {
-      return SymNode(
-          c10::make_intrusive<ConstantSymNodeImpl<bool>>(val_ == *mb_si));
+    auto mb_si = other->singleton_int();
```
You should seriously consider just defining these in terms of their negations, reduce boilerplate

soulitzer replied:
This seems tricky to do because there are a lot of exceptions to the rule. 

| Operation | `< 2`  | `2` | `> 2 but < int64_t::max` | `int64_t::max` |
|-----------|-------|-----|--------------------------|----------------|
| `ge`      | `true`| `true` | Error (Indeterminate) | Error (Indeterminate) |
| `le`      | `false`| Error (Indeterminate) | Error (Indeterminate) | `true` |

soulitzer replied:
With the semantics as-is there are too many edge cases for this to actually be very simplifying:

| Operation | `< 2`  | `2` | `> 2 but < int64_t::max` | `int64_t::max` |
|-----------|-------|-----|--------------------------|----------------|
| `ge`      | `true`| `true` | Indeterminate | Indeterminate | 
| `le`      | `false`| Indeterminate | Indeterminate | `true` |

However, it might actually be reasonable just to make `j <= 2` be `false` and `j > 2` be `true`.

soulitzer replied:
Ignore my previous message, it is indeed possible to simplifying as-is.

soulitzer committed:
Update on "Update singleton int to error when inequality relation is undefined"


Previously, something like j0 >= 3, would return False. In sympy however, it is not possible to make it so that both j0 >= 3 and j0 < 3 return False. In sympy, you only get to dispatch on Ge, and the remaining are derived, e.g. defining Ge(j0 >= 3) to be False would force Lt(j0, 3) to be True, which is not what we want. 

In this PR, we make it so that both j0 >=3 and j0 < 3 error, so that in a future PR when we create the symbolic counterpart of this singleton, the behaviors can be the same.

[ghstack-poisoned]

soulitzer committed:
Update on "Update singleton int to error when inequality relation is undefined"


Previously, something like j0 >= 3, would return False. In sympy however, it is not possible to make it so that both j0 >= 3 and j0 < 3 return False. In sympy, you only get to dispatch on Ge, and the remaining are derived, e.g. defining Ge(j0 >= 3) to be False would force Lt(j0, 3) to be True, which is not what we want. 

In this PR, we make it so that both j0 >=3 and j0 < 3 error, so that in a future PR when we create the symbolic counterpart of this singleton, the behaviors can be the same.

[ghstack-poisoned]

PaliC commented:
@pytorchbot revert -m "bottom diff is causing a plethora of internal failures" -c ghfirst
